databuilder/backends/tpp.py

```diff
@@ -257,6 +257,101 @@ class TPPBackend(BaseBackend):
+                    ELSE 'Unknown'
+                END AS renal_mhyn,
+                CASE
+                    WHEN mildliver = 'YES' THEN 'Yes'
+                    WHEN mildliver IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS mildliver,
+                CASE
+                    WHEN modliv = 'YES' THEN 'Yes'
+                    WHEN modliv IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS modliv,
+                CASE
+                    WHEN chronicneu_mhyn = 'YES' THEN 'Yes'
+                    WHEN chronicneu_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS chronicneu_mhyn,
+                CASE
+                    WHEN malignantneo_mhyn = 'YES' THEN 'Yes'
+                    WHEN malignantneo_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS malignantneo_mhyn,
+                CASE
+                    WHEN chronichaemo_mhyn = 'YES' THEN 'Yes'
+                    WHEN chronichaemo_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS chronichaemo_mhyn,
+                CASE
+                    WHEN aidshiv_mhyn = 'YES' THEN 'Yes'
+                    WHEN aidshiv_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS aidshiv_mhyn,
+                CASE
+                    WHEN obesity_mhyn = 'YES' THEN 'Yes'
+                    WHEN obesity_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS obesity_mhyn,
+                CASE
+                    WHEN diabetes_type_mhyn IN ('1', '2') THEN diabetes_type_mhyn
+                    WHEN diabetes_type_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS diabetes_type_mhyn,
+                CASE
+                    WHEN diabetescom_mhyn = 'YES' THEN 'Yes'
+                    WHEN diabetescom_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS diabetescom_mhyn,
+                CASE
+                    WHEN diabetes_mhyn = 'YES' THEN 'Yes'
+                    WHEN diabetes_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS diabetes_mhyn,
+                CASE
+                    WHEN rheumatologic_mhyn = 'YES' THEN 'Yes'
+                    WHEN rheumatologic_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS rheumatologic_mhyn,
+                CASE
+                    WHEN dementia_mhyn = 'YES' THEN 'Yes'
+                    WHEN dementia_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS dementia_mhyn,
+                CASE
+                    WHEN malnutrition_mhyn = 'YES' THEN 'Yes'
+                    WHEN malnutrition_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
```

Comment:
question: The `ELSE` is setting the value to `Unknown`, because that's the only string that should be left as a column value are matching.

Is it preferable to check explicitly, and result in a `null` if no match at all?

Reply:
Would a researcher ever wish to distinguish "no data" (`null`) from "not known" (`Unknown`)? I don't think they would, so I think this is the correct implementation.

Reply:
I'm not sure this is right, following @wjchulme's comment about [structural nulls](https://github.com/opensafely-core/databuilder/issues/1071#issuecomment-1441450857). I think _either_ NULL means something different from `Unknown`, in which case the logic above needs to distinguish them. Or, if it doesn't mean anything different, then these should be modelled as nullable booleans rather than three-valued categoricals.

I don't have a view as to which is the right answer here. But I think it needs to be one or the other and the status quo sits halfway between them.

Reply:
Ah, I'd not seen Will's comment. I'd also not considered nullable booleans as an alternative to three-valued categoricals. Thanks, Dave 👍🏻.

Reply:
We should use [the data dictionary](https://raw.githubusercontent.com/isaric4c/wiki/master/ISARIC/CCP_REDCap_ISARIC_data_dictionary_codebook.pdf) to guide us here. 

If the "cormob_none" button is clicked, then none of the remaining comorbidity fields are shown, and all will take value `"NA"` in the table. (I think originally this would have been represented as `null` in redcap, but then I think the data went to R, and then it went as a csv to tpp and ended up as `"NA"`). Otherwise, `comorb_none` is not clicked ( -> `null` -> `"NA"`), and so all other comorbidity fields are shown and _must_ be answered (becasue they are "required" fields).

`"Unknown"` in this case means "the question has been considered and the answer is 'we don't know'". This is distinct from "the question has not been considered or has been left unanswered". Iain is probably right that, epistemologically, this makes no difference to researchers, but I still think an explicit `"Unknown"` category here is helpful as it removes all ambiguity. It also distinguishes it from " code is present / code is absent" booleans derived from coded event data, where missingness/unknownness is not a meaningful concept.

Reply:
OK, then I think the answer is that we need to explicitly handle "Unknown" as a case and let the default be NULL. Something like:
```
  WHEN malnutrition_mhyn = 'Unknown' THEN 'Unknown'
END AS malnutrition_mhyn
```

And then the ehrQL columns should become three valued categoricals.